include/hipSYCL/sycl/handler.hpp

```diff
@@ -105,20 +106,22 @@ class handler {
           "because it is not bound to a buffer."};
     }
 
-    size_t element_size = data.mem->get_element_size();
-
-    if (element_size != sizeof(typename AccessorType::value_type)) {
-      assert(false && "Reinterpreting data with elements of different size is "
-                      "not yet supported");
-    }
-
     // Translate no_init property and host_task modes
     access_mode mode =
         detail::get_effective_access_mode(AccessorType::mode, data.is_no_init);
+    
+    size_t element_size = data.mem->get_element_size();
 
-    auto req = std::make_unique<rt::buffer_memory_requirement>(
+    std::unique_ptr<rt::buffer_memory_requirement> req;
+    if (element_size != sizeof(typename AccessorType::value_type)) {
+      req.reset(new rt::buffer_memory_requirement(
+        data.mem, rt::id<3>{}, data.mem->get_num_elements(), mode,
+        AccessorType::access_target));
+    } else {
+      req.reset(new rt::buffer_memory_requirement(
         data.mem, rt::make_id(data.offset), rt::make_range(data.range), mode,
-        AccessorType::access_target);
+        AccessorType::access_target));
```

Comment:
We have this logic now in multiple places. I wonder if it might be a good idea to have a single `get_effective_offset/range` or similar somewhere, so that we only need to change a single place if we want to optimize this case some more?